### PR TITLE
[PyTorch] TransformerLayer: add support for Falcon architecture

### PR DESCRIPTION
**Pull request by Marks101**
[Falcon-40b](https://huggingface.co/tiiuae/falcon-40b) and [Falcon-180b](https://huggingface.co/tiiuae/falcon-180B) are two exciting publicly available models. Currently, transformer-engine does not support their architecture because in their implementation self attention and mlp are not computed in sequence. Instead, the blocks (layer norm -> self attention) and (layernorm -> mlp) are fed with the input into the layer. In a computational graph these operations are thus in parallel. In the Falcon configs this is denoted as `new_decoder_architecture`. This PR introduces this feature and thus makes it possible to finetune Falcon models with transformer-engine. We would be really happy if this feature finds it's way into transformer-engine.

Two notes on the implementation:
- I denoted the new option `parallel_attention_mlp`. I am not sure if this is a perfect naming, this is up for discussion.
- I created a new method `_bias_dropout_add()` in order to keep the new code clear
- Falcon models do not have a bias, accordingly I think that a pretty clean solution is to set `return_bias=False` for attention and mlp and then use bias dropout add. For models using `parallel_attention_mpl` and `bias`, this might not be optimal.

### CODE REVIEW

**Review by timmoon10 — APPROVED** [review]
LGTM

**Comment by Marks101** [comment]
Thanks for proceeding so quickly with this PR 🥳

**Review by ptrendx — APPROVED** [review]

File: tests/pytorch/test_numerics.py [hunk 1]
```diff
@@ -318,13 +318,12 @@ def forward(self, x):
 
 
 class TorchGPT(nn.Module):
-    def __init__(self, hidden_size: int, eps: float, num_attention_heads: int):
+    def __init__(self, hidden_size: int, eps: float, num_attention_heads: int, parallel_attention_ml: bool):
```
Comment (timmoon10) [inline tests/pytorch/test_numerics.py:321]:
```suggestion
    def __init__(self, hidden_size: int, eps: float, num_attention_heads: int, parallel_attention_mlp: bool):
```

File: tests/pytorch/test_numerics.py [hunk 2]
```diff
@@ -318,13 +318,12 @@ def forward(self, x):
...
         super().__init__()
         self.ln = nn.LayerNorm(hidden_size, eps=eps)
         self.causal_attn = TorchMHA(hidden_size, num_attention_heads)
         self.ln_mlp = TorchLayerNormMLP(hidden_size, 4 * hidden_size, eps)
-        self.resid_attn_dropout = nn.Dropout(0.1)
-        self.resid_mlp_dropout = nn.Dropout(0.1)
+        self.parallel_attention_mlp = parallel_attention_ml
```
Comment (timmoon10) [inline tests/pytorch/test_numerics.py:326]:
```suggestion
        self.parallel_attention_mlp = parallel_attention_mlp
```

File: tests/pytorch/test_numerics.py [hunk 3]
```diff
@@ -619,7 +623,8 @@ def _test_e2e_gpt_accuracy(block, bs, dtype, config):
 @pytest.mark.parametrize("dtype", param_types)
 @pytest.mark.parametrize("bs", batch_sizes)
 @pytest.mark.parametrize("model", model_configs.keys())
-def test_gpt_accuracy(dtype, bs, model):
+@pytest.mark.parametrize("parallel_attention_ml", all_boolean)
+def test_gpt_accuracy(dtype, bs, model, parallel_attention_ml):
```
Comment (timmoon10) [inline tests/pytorch/test_numerics.py:627]:
```suggestion
@pytest.mark.parametrize("parallel_attention_mlp", all_boolean)
def test_gpt_accuracy(dtype, bs, model, parallel_attention_mlp):
```

File: tests/pytorch/test_numerics.py [hunk 4]
```diff
@@ -632,6 +637,7 @@ def test_gpt_accuracy(dtype, bs, model):
             hidden_dropout=0.1,
             fuse_qkv_params=True,
             qkv_weight_interleaved=False,
+            parallel_attention_mlp=parallel_attention_ml,
```
Comment (timmoon10) [inline tests/pytorch/test_numerics.py:640]:
```suggestion
            parallel_attention_mlp=parallel_attention_mlp,
```

File: tests/pytorch/test_numerics.py [hunk 5]
```diff
@@ -643,6 +649,7 @@ def test_gpt_accuracy(dtype, bs, model):
             config.hidden_size,
             config.eps,
             config.num_attention_heads,
+            parallel_attention_ml=parallel_attention_ml,
```
Comment (timmoon10) [inline tests/pytorch/test_numerics.py:652]:
```suggestion
            parallel_attention_ml=parallel_attention_mlp,
```

**Comment by Marks101** [comment]
Thanks for fixing my spelling mistakes ... sorry for that. The tests failed because there was one last occurance of `parallel_attention_ml`. I fixed that and ran the tests locally. Now it should be fine.

**Comment by ptrendx** [comment]
Tim's attempt to run the CI failed due to network issue apparently, I just retried it.

**Comment by ptrendx** [comment]
Merged. Thank you @Marks101 for the contribution!

**Comment by Marks101** [comment]
Great, thank you for the support!
